chore!: Update versions

joriscode opened the pull request:
- Update Scala, Sbt, JDK, librairies, plugins versions
- Drop Java 8 support
- Remove Scalastyle
- Add Scalafix
- Update scalafmt.conf
- Apply scalafix and scalafmt
- Fix compilation errors due to libraries updates
- Update cli.yml
  - Skip duplicated workflows
  - Update Scala, JDK and Github actions versions
- Add Scala Steward badge
- Bump release version to 2.9.0-SNAPSHOT

joriscode commented on build.sbt:
```diff
@@ -1,11 +1,16 @@
+// Enable scalafix semantic rules
```
I'm moving this to `BuildPlugin`

nrinaudo commented on core/src/main/scala/kantan/sbt/KantanBoilerplatePlugin.scala:
```diff
@@ -30,15 +32,15 @@ object KantanBoilerplatePlugin extends AutoPlugin {
   override def trigger = allRequirements
 
-  override def requires = HeaderPlugin && BoilerplatePlugin
+  override def requires: Plugins = HeaderPlugin && BoilerplatePlugin
```
Note: I would never block a PR for that, but I personally never ascribe a type to overriden members unless I'm narrowing it. Scala is syntax heavy as i is, so I like to remove anything that doesn't bring new information.

joriscode replied:
I agree with you. Unfortunately, this is due to scalafix `ExplicitResultTypes` rule (https://scalacenter.github.io/scalafix/docs/rules/ExplicitResultTypes.html) and there is no setting to prevent this, except disabling Scalafix on these lines.
Note: `override def trigger = allRequirements` two lines above is not edited since the default value of the rule exclude "simple definitions" (`skipSimpleDefinitions`).

nrinaudo commented on kantan/src/main/resources/kantan/sbt/scalafix.conf:
```diff
@@ -0,0 +1,16 @@
+  # Semantic Rules
+  ExplicitResultTypes, # Inserts type annotations for inferred public members.
+  NoAutoTupling, # Inserts explicit tuples for adapted argument lists for compatibility with -Yno-adapted-args
+  OrganizeImports, # Organize import statements
```
Oh interesting, I'll need to look up how that works - I thought it was essentially an impossible task because the order of imports changes the meaning of the code!

Or does this do the same thing as scalafmt - simply sort imports in an `import foo.{A, B, C}` statement?

joriscode replied:
Here is the documentation of this rule: https://scalacenter.github.io/scalafix/docs/rules/OrganizeImports.html
To answer your question, it does re-order import lines. It can do more, such as expanding partially-qualified imports to fully-qualified.

nrinaudo commented on kantan/src/main/scala/kantan/sbt/kantan/KantanAutomateHeaderPlugin.scala:
```diff
@@ -16,12 +16,13 @@
 object KantanAutomateHeaderPlugin extends AutoPlugin {
-  override def trigger  = allRequirements
-  override def requires = KantanKantanPlugin && HeaderPlugin
+  override def trigger           = allRequirements
```
Huh. This is not very consistent. Is that scalafix failing to add this type annotation when it added one for the next line?

nrinaudo commented:
I'm pretty sure this is a preparation PR for a bunch more, probably on more interesting projects that depend on this one.

For these ulterior PRs, could I ask you to separate at least the commit in which you apply the new scalafix / scalafmt rules and the commit(s) in which you do the interesting work? It would make them much easier to review, not having to wonder for every diff if it's merely style or if it has substance.

joriscode commented on .github/workflows/cli.yml:
```diff
@@ -1,30 +1,42 @@
       matrix:
         os   : [ubuntu-latest]
-        scala: [2.12.15]
-        java : [8]
```
I bumped into issues with Java 8 on _Kantan codecs_ and decided to drop it altogether

joriscode replied:
Would you prefer to keep Java 8 support in this project as long as it is possible even if other projects (codecs, csv) cannot maintain the support?

joriscode commented on .gitignore:
```diff
@@ -1,36 +1,9 @@
-.DS_Store
-
-# Auto-generated files
-.scalafmt.conf
```
`.scalafmt.conf` (resp. `.scalafix.conf`) here is a symlink to the config defined in the plugin `kantan/src/main/resources/kantan/sbt/scalafmt.conf`.
As the build of project cannot depend on `ScalafmtPlugin`, `.scalafmt.conf` is not generated.

joriscode commented:
I force-pushed to separate the original commit in three (changes, format, bump release version).
I changed formatting rules:
- removed Scalafmt `ImportSorts`
- set Scalafix rule to enforce trailing-commas (reduces diffs on the long run)
- set Scalafix rule to expand imports to fully-qualified format (without that Scalafix order would have them in a separate block below the other imports)

joriscode commented:
Force-pushed:
- Reverse removal of Java 8 support
- Downgrade of dependencies preventing Java 8 support
- Rename `BuildPlugin` to `LocalBuildPlugin`

joriscode commented:
Successful workflow on my repo: https://github.com/joriscode/kantan.sbt/actions/runs/12699906331

joriscode commented:
Force-pushed again:
- Update mdoc, scalafmt versions to latest while ignoring `com.github.plokhotnyuk.jsoniter-scala` transitive dependency

joriscode commented:
Force-pushed:
- Update scalajs to `1.18.1`

nrinaudo commented on build.sbt:
```diff
@@ -3,15 +3,15 @@ lazy val root = Project(id = "kantan-sbt", base = file("."))
     publish         := {},
     publishLocal    := {},
-    publishArtifact := false
+    publishArtifact := false,
```
Is there any chance we could revert this change? I really, really do not like dangling commas...

joriscode commented:
Force-push:
- Revert trailing commas setting
- Update `releaseStepCommand("scalafix")` to `releaseStepCommand("scalafix --check")`
